File: server/src/main/java/dev/slimevr/poserecorder/PoseFrameIO.java
package dev.slimevr.poserecorder;

import com.jme3.math.Quaternion;
import com.jme3.math.Vector3f;
import dev.slimevr.tracking.trackers.TrackerPosition;
import io.eiren.util.collections.FastList;
import io.eiren.util.logging.LogManager;

import java.io.*;


public final class PoseFrameIO {

	private PoseFrameIO() {
		// Do not allow instantiating
	}

	private static void writeVector3f(DataOutputStream outputStream, Vector3f vector)
		throws IOException {
		outputStream.writeFloat(vector.getX());
		outputStream.writeFloat(vector.getY());
		outputStream.writeFloat(vector.getZ());
	}

	private static void writeQuaternion(DataOutputStream outputStream, Quaternion quaternion)
		throws IOException {
		outputStream.writeFloat(quaternion.getX());
		outputStream.writeFloat(quaternion.getY());
		outputStream.writeFloat(quaternion.getZ());
		outputStream.writeFloat(quaternion.getW());
	}

	public static boolean writeFrames(DataOutputStream outputStream, PoseFrames frames) {
		try {
			if (frames != null) {
				outputStream.writeInt(frames.getTrackerCount());
				for (PoseFrameTracker tracker : frames.getTrackers()) {
					outputStream.writeUTF(tracker.name);
					outputStream.writeInt(tracker.getFrameCount());
					for (int i = 0; i < tracker.getFrameCount(); i++) {
						TrackerFrame trackerFrame = tracker.safeGetFrame(i);
						if (trackerFrame == null) {
							outputStream.writeInt(0);
							continue;
						}

						int dataFlags = trackerFrame.getDataFlags();

						// Don't write destination strings anymore, replace with
						// the enum
						if (trackerFrame.hasData(TrackerFrameData.DESIGNATION_STRING)) {
							dataFlags = TrackerFrameData.DESIGNATION_ENUM
								.add(TrackerFrameData.DESIGNATION_STRING.remove(dataFlags));
						}

						outputStream.writeInt(dataFlags);

						if (trackerFrame.hasData(TrackerFrameData.ROTATION)) {
							writeQuaternion(outputStream, trackerFrame.rotation);
						}

						if (trackerFrame.hasData(TrackerFrameData.POSITION)) {
							writeVector3f(outputStream, trackerFrame.position);
						}

						if (TrackerFrameData.DESIGNATION_ENUM.check(dataFlags)) {
							outputStream.writeInt(trackerFrame.designation.ordinal());
						}

						if (trackerFrame.hasData(TrackerFrameData.ACCELERATION)) {
							writeVector3f(outputStream, trackerFrame.acceleration);
						}

						if (trackerFrame.hasData(TrackerFrameData.RAW_ROTATION)) {
							writeQuaternion(outputStream, trackerFrame.rawRotation);
						}
					}
				}
			} else {
				outputStream.writeInt(0);
			}
		} catch (Exception e) {
			LogManager.severe("Error writing frame to stream", e);
			return false;
		}

		return true;
	}

	public static boolean writeToFile(File file, PoseFrames frames) {
		try (
			DataOutputStream outputStream = new DataOutputStream(
				new BufferedOutputStream(new FileOutputStream(file))
			)
		) {
			writeFrames(outputStream, frames);
		} catch (Exception e) {
			LogManager.severe("Error writing frames to file", e);
			return false;
		}

		return true;
	}

	private static Vector3f readVector3f(DataInputStream inputStream) throws IOException {
		return new Vector3f(
			inputStream.readFloat(),
			inputStream.readFloat(),
			inputStream.readFloat()
		);
	}

	private static Quaternion readQuaternion(DataInputStream inputStream) throws IOException {
		return new Quaternion(
			inputStream.readFloat(),
			inputStream.readFloat(),
			inputStream.readFloat(),
			inputStream.readFloat()
		);
	}

	public static PoseFrames readFrames(DataInputStream inputStream) {
		try {

			int trackerCount = inputStream.readInt();
			FastList<PoseFrameTracker> trackers = new FastList<>(trackerCount);
			for (int i = 0; i < trackerCount; i++) {

				String name = inputStream.readUTF();
				int trackerFrameCount = inputStream.readInt();
				FastList<TrackerFrame> trackerFrames = new FastList<>(
					trackerFrameCount
				);
				for (int j = 0; j < trackerFrameCount; j++) {
					int dataFlags = inputStream.readInt();

					TrackerPosition designation = null;
					if (TrackerFrameData.DESIGNATION_STRING.check(dataFlags)) {
						designation = TrackerPosition
							.getByDesignation(inputStream.readUTF())
							.orElse(null);
					}

					Quaternion rotation = null;
					if (TrackerFrameData.ROTATION.check(dataFlags)) {
						rotation = readQuaternion(inputStream);
					}

					Vector3f position = null;
					if (TrackerFrameData.POSITION.check(dataFlags)) {
						position = readVector3f(inputStream);
					}

					if (TrackerFrameData.DESIGNATION_ENUM.check(dataFlags)) {
						designation = TrackerPosition.values[inputStream.readInt()];
					}

					Vector3f acceleration = null;
					if (TrackerFrameData.ACCELERATION.check(dataFlags)) {
						acceleration = readVector3f(inputStream);
					}

					Quaternion rawRotation = null;
					if (TrackerFrameData.RAW_ROTATION.check(dataFlags)) {
						rawRotation = readQuaternion(inputStream);
					}

					trackerFrames
						.add(
							new TrackerFrame(
								designation,
								rotation,
								position,
								acceleration,
								rawRotation
							)
						);
				}

				trackers.add(new PoseFrameTracker(name, trackerFrames));
			}

			return new PoseFrames(trackers);
		} catch (Exception e) {
			LogManager.severe("Error reading frame from stream", e);
		}

		return null;
	}

	public static PoseFrames readFromFile(File file) {
		try {
			return readFrames(
				new DataInputStream(new BufferedInputStream(new FileInputStream(file)))
			);
		} catch (Exception e) {
			LogManager.severe("Error reading frame from file", e);
		}

		return null;
	}
}

File: server/src/main/java/dev/slimevr/poseframeformat/PoseFrameIO.kt
package dev.slimevr.poseframeformat

import dev.slimevr.poseframeformat.trackerdata.TrackerFrame
import dev.slimevr.poseframeformat.trackerdata.TrackerFrameData
import dev.slimevr.poseframeformat.trackerdata.TrackerFrames
import dev.slimevr.tracking.trackers.TrackerPosition
import dev.slimevr.tracking.trackers.TrackerPosition.Companion.getByDesignation
import io.eiren.util.collections.FastList
import io.eiren.util.logging.LogManager
import io.github.axisangles.ktmath.Quaternion
import io.github.axisangles.ktmath.Vector3
import java.io.BufferedInputStream
import java.io.BufferedOutputStream
import java.io.DataInputStream
import java.io.DataOutputStream
import java.io.File
import java.io.FileInputStream
import java.io.FileOutputStream
import java.io.IOException

object PoseFrameIO {
	@Throws(IOException::class)
	private fun writeVector3f(outputStream: DataOutputStream, vector: Vector3) {
		outputStream.writeFloat(vector.x)
		outputStream.writeFloat(vector.y)
		outputStream.writeFloat(vector.z)
	}

	@Throws(IOException::class)
	private fun writeQuaternion(outputStream: DataOutputStream, quaternion: Quaternion) {
		outputStream.writeFloat(quaternion.x)
		outputStream.writeFloat(quaternion.y)
		outputStream.writeFloat(quaternion.z)
		outputStream.writeFloat(quaternion.w)
	}

	fun writeFrames(outputStream: DataOutputStream, frames: PoseFrames) {
		outputStream.writeInt(frames.frameHolders.size)
		for (tracker in frames.frameHolders) {
			outputStream.writeUTF(tracker.name)
			outputStream.writeInt(tracker.frames.size)
			for (i in 0 until tracker.frames.size) {
				val trackerFrame = tracker.tryGetFrame(i)
				if (trackerFrame == null) {
					outputStream.writeInt(0)
					continue
				}
				var dataFlags = trackerFrame.dataFlags

				// Don't write destination strings anymore, replace with
				// the enum
				if (trackerFrame.hasData(TrackerFrameData.DESIGNATION_STRING)) {
					dataFlags = TrackerFrameData.TRACKER_POSITION_ENUM
						.add(TrackerFrameData.DESIGNATION_STRING.remove(dataFlags))
				}
				outputStream.writeInt(dataFlags)
				if (trackerFrame.hasData(TrackerFrameData.ROTATION)) {
					writeQuaternion(outputStream, trackerFrame.rotation!!)
				}
				if (trackerFrame.hasData(TrackerFrameData.POSITION)) {
					writeVector3f(outputStream, trackerFrame.position!!)
				}
				if (TrackerFrameData.TRACKER_POSITION_ENUM.check(dataFlags)) {
					outputStream.writeInt(trackerFrame.trackerPosition!!.ordinal)
				}
				if (trackerFrame.hasData(TrackerFrameData.ACCELERATION)) {
					writeVector3f(outputStream, trackerFrame.acceleration!!)
				}
				if (trackerFrame.hasData(TrackerFrameData.RAW_ROTATION)) {
					writeQuaternion(outputStream, trackerFrame.rawRotation!!)
				}
			}
		}
	}

	fun tryWriteFrames(outputStream: DataOutputStream, frames: PoseFrames): Boolean {
		try {
			writeFrames(outputStream, frames)
		} catch (e: Exception) {
			LogManager.severe("Error writing frame to stream", e)
			return false
		}
		return true
	}

	fun writeToFile(file: File, frames: PoseFrames) {
		DataOutputStream(
			BufferedOutputStream(FileOutputStream(file))
		).use { outputStream -> writeFrames(outputStream, frames) }
	}

	fun tryWriteToFile(file: File, frames: PoseFrames): Boolean {
		try {
			writeToFile(file, frames)
		} catch (e: Exception) {
			LogManager.severe("Error writing frames to file", e)
			return false
		}
		return true
	}

	@Throws(IOException::class)
	private fun readVector3f(inputStream: DataInputStream): Vector3 {
		return Vector3(
			inputStream.readFloat(),
			inputStream.readFloat(),
			inputStream.readFloat()
		)
	}

	@Throws(IOException::class)
	private fun readQuaternion(inputStream: DataInputStream): Quaternion {
		val x = inputStream.readFloat()
		val y = inputStream.readFloat()
		val z = inputStream.readFloat()
		val w = inputStream.readFloat()

		return Quaternion(w, x, y, z)
	}

	fun readFrames(inputStream: DataInputStream): PoseFrames {
		val trackerCount = inputStream.readInt()
		val trackers = FastList<TrackerFrames>(trackerCount)
		for (i in 0 until trackerCount) {
			val name = inputStream.readUTF()
			val trackerFrameCount = inputStream.readInt()
			val trackerFrames = FastList<TrackerFrame?>(
				trackerFrameCount
			)
			for (j in 0 until trackerFrameCount) {
				val dataFlags = inputStream.readInt()
				var designation: TrackerPosition? = null
				if (TrackerFrameData.DESIGNATION_STRING.check(dataFlags)) {
					designation = getByDesignation(inputStream.readUTF())
				}
				var rotation: Quaternion? = null
				if (TrackerFrameData.ROTATION.check(dataFlags)) {
					rotation = readQuaternion(inputStream)
				}
				var position: Vector3? = null
				if (TrackerFrameData.POSITION.check(dataFlags)) {
					position = readVector3f(inputStream)
				}
				if (TrackerFrameData.TRACKER_POSITION_ENUM.check(dataFlags)) {
					designation = TrackerPosition.values()[inputStream.readInt()]
				}
				var acceleration: Vector3? = null
				if (TrackerFrameData.ACCELERATION.check(dataFlags)) {
					acceleration = readVector3f(inputStream)
				}
				var rawRotation: Quaternion? = null
				if (TrackerFrameData.RAW_ROTATION.check(dataFlags)) {
					rawRotation = readQuaternion(inputStream)
				}
				trackerFrames
					.add(
						TrackerFrame(
							designation,
							rotation,
							position,
							acceleration,
							rawRotation
						)
					)
			}
			trackers.add(TrackerFrames(name, trackerFrames))
		}
		return PoseFrames(trackers)
	}

	fun tryReadFrames(inputStream: DataInputStream): PoseFrames? {
		return try {
			return readFrames(inputStream)
		} catch (e: Exception) {
			LogManager.severe("Error reading frames from stream", e)
			null
		}
	}

	fun readFromFile(file: File): PoseFrames {
		return readFrames(
			DataInputStream(BufferedInputStream(FileInputStream(file)))
		)
	}

	fun tryReadFromFile(file: File): PoseFrames? {
		return try {
			return readFrames(
				DataInputStream(BufferedInputStream(FileInputStream(file)))
			)
		} catch (e: Exception) {
			LogManager.severe("Error reading frames from file", e)
			null
		}
	}
}
